cpp/src/community/ECG.cu

```diff
@@ -0,0 +1,328 @@
+    cugraph::Graph* permuted = permute_graph<IdxT, ValT>(graph, permutation);
+
+    // Run Louvain clustering on the random permutation
+    gdf_column* parts_col = (gdf_column*) malloc(sizeof(gdf_column));
+    IdxT* parts;
+    ALLOC_TRY(&parts, sizeof(IdxT) * size, nullptr);
+    gdf_column_view(parts_col, parts, nullptr, size, graph->adjList->offsets->dtype);
+    ValT final_modularity;
+    IdxT num_level;
+    cugraph::louvain(permuted, &final_modularity, &num_level, parts_col, 1);
+
+    // For each edge in the graph determine whether the endpoints are in the same partition
+    // Keep a sum for each edge of the total number of times its endpoints are in the same partition
+    dim3 grid, block;
+    block.x = 512;
+    grid.x = min((IdxT) MAXBLOCKS, (nnz / 512 + 1));
+    match_check_kernel<<<grid, block, 0, nullptr>>>(nnz,
+                                                    size,
+                                                    offsets,
+                                                    indices,
+                                                    permutation,
+                                                    (IdxT*) parts_col->data,
+                                                    ecg_weights);
+
+    // Clean up temporary allocations
+    delete permuted;
+    ALLOC_FREE_TRY(parts, nullptr);
+    free(parts_col);
+    ALLOC_FREE_TRY(permutation, nullptr);
+  }
+
+  // Set weights = min_weight + (1 - min-weight)*sum/ensemble_size
+  update_functor<ValT> uf(min_weight, ensemble_size);
+  thrust::transform(rmm::exec_policy(nullptr)->on(nullptr), ecg_weights, ecg_weights + nnz, ecg_weights, uf);
+
+  // Run Louvain on the original graph using the computed weights
+  cugraph::Graph* result = new cugraph::Graph;
+  result->adjList = new cugraph::gdf_adj_list;
+  result->adjList->offsets = new gdf_column;
+  result->adjList->indices = new gdf_column;
+  result->adjList->edge_data = new gdf_column;
+  result->adjList->ownership = 0;
+  gdf_column_view(result->adjList->offsets,
+                  offsets,
+                  nullptr,
+                  graph->adjList->offsets->size,
+                  graph->adjList->offsets->dtype);
+  gdf_column_view(result->adjList->indices,
+                  indices,
+                  nullptr,
+                  graph->adjList->indices->size,
+                  graph->adjList->indices->dtype);
+  gdf_column_view(result->adjList->edge_data,
+                  ecg_weights,
+                  nullptr,
+                  graph->adjList->edge_data->size,
+                  graph->adjList->edge_data->dtype);
+  ValT final_modularity;
+  IdxT num_level;
+  cugraph::louvain(result, &final_modularity, &num_level, ecg_parts, 100);
+
+  // Cleaning up temporary allocations
+  delete result;
+  ALLOC_FREE_TRY(ecg_weights, nullptr);
+}
+} // anonymous namespace
+
+
+namespace cugraph {
+void ecg(Graph* graph,
+         double min_weight,
+         int ensemble_size,
+         gdf_column *ecg_parts) {
```

Comment:
I have updated the API for ECG and Louvain to not use gdf_column.

Reply:
Thanks. While there is no `gdf_column` anymore it is not yet consistent with the new API. The function should be templated instead of using a `void` pointer for instance. Something like this:
```
/**                                                                             
 *     ...
 * @tparam VT the type of vertex identifiers. Supported value : int (signed, 32-bit)
 * @tparam WT the type of edge weights. Supported value : float or double.                                                      
 *     ...
 */                                                                             
/* ----------------------------------------------------------------------------*/
template <typename VT, typename WT>
void algo (Graph* graph,                                            
          WT *some_weights_type_set,                                                  
          VT *some_vertex_type_set, ...);                      
```

Should `ensemble_size` be `size_t` instead of `int` or is there a use case for negative values?

Is `min_weight` compared to something of type `WT` at any time? If so, `WT` it may be better if it is a `WT` too.

Reply:
Updated.

Reply:
umm I don't see any new commits at this point. It says `Latest commit 2b9a9ea 2 days ago`